// vnpay/lib/enums/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RefundTransactionType = exports.VnpTransactionType = exports.VnpLocale = exports.VnpCurrCode = exports.HashAlgorithm = exports.UrlService = void 0;
exports.UrlService = {
    sandbox: 'https://sandbox.vnpayment.vn/paymentv2/vpcpay.html',
};
exports.HashAlgorithm = {
    SHA256: 'SHA256',
    SHA512: 'SHA512',
    MD5: 'MD5',
};
exports.VnpCurrCode = {
    VND: 'VND',
};
exports.VnpLocale = {
    VN: 'vn',
    EN: 'en',
};
var VnpTransactionType;
(function (VnpTransactionType) {
    VnpTransactionType["PAYMENT"] = "01";
    VnpTransactionType["FULL_REFUND"] = "02";
    VnpTransactionType["PARTIAL_REFUND"] = "03";
})(VnpTransactionType || (exports.VnpTransactionType = VnpTransactionType = {}));
var RefundTransactionType;
(function (RefundTransactionType) {
    RefundTransactionType["FULL_REFUND"] = "02";
    RefundTransactionType["PARTIAL_REFUND"] = "03";
})(RefundTransactionType || (exports.RefundTransactionType = RefundTransactionType = {}));
